Circle cache improvements

alan-agius4 opened the pull request:
**ci: add cache for windows run**
Add CircleCI cache for windows runs


**ci: don't save node_modules directory**
Storing node_modules directory in cache causes a number of issues, such as increase in the laid out node_modules tree and also in some cases break module resolution. This is because modules are not hoisted correctly when running `yarn install`.
    
Example of related failures https://app.circleci.com/pipelines/github/angular/angular/30441/workflows/c1a16135-f5c9-478d-a298-d720c6657d52/jobs/952138
    
A better cache strategy would be to save yarn cache, more information can be found in https://circleci.com/docs/2.0/caching/#yarn-node.

alan-agius4 committed:
ci: add cache for windows run

Add CircleCI cache for windows runs

gkalpak commented on .circleci/config.yml:
```diff
@@ -22,11 +22,11 @@ version: 2.1
 # **NOTE 2 **: Keep the static part of the cache key as prefix to enable correct fallbacks.
 # See https://circleci.com/docs/2.0/caching/#restoring-cache for how prefixes work in CircleCI.
-var_3: &cache_key v1-angular-node-12-{{ checksum ".bazelversion" }}-{{ checksum "yarn.lock" }}-{{ checksum "WORKSPACE" }}-{{ checksum "packages/bazel/package.bzl" }}-{{ checksum "aio/yarn.lock" }}
+var_3: &cache_key v3-angular-node-12-{{ checksum ".bazelversion" }}-{{ checksum "yarn.lock" }}-{{ checksum "WORKSPACE" }}-{{ checksum "packages/bazel/package.bzl" }}-{{ checksum "aio/yarn.lock" }}
```
Why `v3` (and not `v2`)? 🤔

alan-agius4 replied:
I used them in a previous PR.

alan-agius4 replied:
Note: not using `-{{ checksum "yarn.lock" }}` as part of the fallback will cause that the cache size to keep growing.

gkalpak commented on .circleci/config.yml:
```diff
@@ -187,7 +187,7 @@ commands:
       - run: node --version
       - run: yarn --version
-      - run: yarn install --frozen-lockfile --non-interactive
+      - run: yarn install --frozen-lockfile --non-interactive --cache-folder ~.cache/yarn
```
Why not `~/.cache-yarn` (here and below)?

```suggestion
      - run: yarn install --frozen-lockfile --non-interactive --cache-folder ~/.cache/yarn
```

gkalpak commented on .circleci/config.yml:
```diff
@@ -772,6 +781,11 @@ jobs:
+      - save_cache:
+          key: *cache_key_win
+          paths:
+            - ~/.cache/yarn
```
Does `~` work on Windows? 🧐

alan-agius4 replied:
In here it does.

gkalpak commented on .circleci/config.yml:
```diff
@@ -703,7 +712,7 @@ jobs:
           # minimum dependency on NodeJS v12 and Yarn 1.19.1, but the framework repository uses
           # older versions.
-          command: yarn --ignore-engines --cwd ${COMPONENTS_REPO_TMP_DIR} install --frozen-lockfile --non-interactive --cache-folder ~.cache/yarn
+          command: yarn --ignore-engines --cwd ${COMPONENTS_REPO_TMP_DIR} install --frozen-lockfile --non-interactive --cache-folder ~/.cache/yarn
```
This change belongs to the previous commit.

gkalpak commented on .circleci/config.yml:
```diff
@@ -232,10 +241,10 @@ jobs:
           # Yarn's requests sometimes take more than 10mins to complete.
           no_output_timeout: 45m
-      - run: yarn --cwd aio install --frozen-lockfile --non-interactive --cache-folder ~.cache/yarn
+      - run: yarn --cwd aio install --frozen-lockfile --non-interactive --cache-folder ~/.cache/yarn
```
This change belongs to the previous commit.

gkalpak commented on .circleci/config.yml:
```diff
@@ -232,10 +241,10 @@ jobs:
       - run:
           name: Running Yarn install
-          command: yarn install --frozen-lockfile --non-interactive --cache-folder ~.cache/yarn
+          command: yarn install --frozen-lockfile --non-interactive --cache-folder ~/.cache/yarn
```
This change belongs to the previous commit.

gkalpak commented on .circleci/config.yml:
```diff
@@ -772,6 +781,11 @@ jobs:
           command: yarn bazel test --config=ivy --build_tag_filters=-no-windows --test_tag_filters="-no-ivy-aot,-no-windows,-fixme-ivy-aot,-browser:chromium-local" //packages/compiler-cli/... //tools/ts-api-guardian/... //packages/localize/...
           no_output_timeout: 15m
+      # Save dependencies to use on subsequent runs.
+      - save_cache:
```
Shouldn't this be done on `setup_win`?

gkalpak commented on .circleci/config.yml:
```diff
@@ -187,7 +191,12 @@ commands:
+          keys:
+            - *cache_key_win
+            - *cache_key_win_fallback
+      # On Windows `~/` is not resolved when using as a CLI value. `../` results in the same path.
```
Actually, the Windows jobs use Powershell, which should support `~`.
(Otherwise, we could use the `userprofile` env variable - e.g. `%userprofile%` in CMD or `$Env:userprofile` in PowerShell.)

gkalpak replied:
I just checked by SSH'ing into the container and it worked fine (when using PowerShell) 🤷‍♂️

alan-agius4 replied:
```
PS C:\Users\circleci\ng> ls
ls


    Directory: C:\Users\circleci\ng


Mode                LastWriteTime         Length Name
----                -------------         ------ ----
d-----         4/6/2021   1:46 PM                .circleci
d-----         4/6/2021   1:46 PM                .devcontainer
d-----         4/6/2021   1:46 PM                .github
d-----         4/6/2021   1:46 PM                .husky
d-----         4/6/2021   1:46 PM                .ng-dev
d-----         4/6/2021   1:46 PM                .vscode
d-----         4/6/2021   1:46 PM                .yarn
d-----         4/6/2021   1:46 PM                aio
d-----         4/6/2021   1:46 PM                dev-infra
d-----         4/6/2021   1:46 PM                docs
d-----         4/6/2021   1:46 PM                goldens
d-----         4/6/2021   1:46 PM                integration
d-----         4/6/2021   1:46 PM                modules
d-----         4/6/2021   1:46 PM                packages
d-----         4/6/2021   1:46 PM                scripts
d-----         4/6/2021   1:46 PM                third_party
d-----         4/6/2021   1:46 PM                tools
d-----         4/6/2021   1:48 PM                ~
-a----         4/6/2021   1:46 PM           4694 .bazelignore
-a----         4/6/2021   1:46 PM           6938 .bazelrc
-a----         4/6/2021   1:46 PM            216 .bazelversion
-a----         4/6/2021   1:46 PM             76 .clang-format
-a----         4/6/2021   1:46 PM            260 .editorconfig
-a----         4/6/2021   1:46 PM            326 .gitattributes
-a----         4/6/2021   1:46 PM            900 .gitignore
-a----         4/6/2021   1:46 PM           7516 .gitmessage
-a----         4/6/2021   1:46 PM             52 .mailmap
-a----         4/6/2021   1:46 PM              9 .nvmrc
-a----         4/6/2021   1:46 PM          49642 .pullapprove.yml
-a----         4/6/2021   1:46 PM            135 .yarnrc
-a----         4/6/2021   1:46 PM           8575 browser-providers.conf.js
-a----         4/6/2021   1:46 PM           2067 BUILD.bazel
-a----         4/6/2021   1:46 PM         876769 CHANGELOG.md
-a----         4/6/2021   1:46 PM           1174 CODE_OF_CONDUCT.md
-a----         4/6/2021   1:46 PM          16300 CONTRIBUTING.md
-a----         4/6/2021   1:46 PM            859 gulpfile.js
-a----         4/6/2021   1:46 PM           8201 karma-js.conf.js
-a----         4/6/2021   1:46 PM           1125 LICENSE
-a----         4/6/2021   1:46 PM           8941 package.json
-a----         4/6/2021   1:46 PM           4923 README.md
-a----         4/6/2021   1:46 PM           1946 renovate.json
-a----         4/6/2021   1:46 PM            259 test-events.js
-a----         4/6/2021   1:46 PM          11640 test-main.js
-a----         4/6/2021   1:46 PM           2613 tslint.json
-a----         4/6/2021   1:46 PM           3564 WORKSPACE
-a----         4/6/2021   1:46 PM         740197 yarn.lock
-a----         4/6/2021   1:46 PM           1356 yarn.lock.readme.md
```

When in circle config we the mentioned line to 

```
- run: yarn install --frozen-lockfile --non-interactive --cache-folder ~/.cache/yarn
```

gkalpak replied:
🤷

gkalpak commented on .circleci/config.yml:
```diff
@@ -22,11 +22,15 @@ version: 2.1
 # folder will contain all previously used versions and ultimately cause the cache restoring to
 # be slower due to its growing size.
-var_4: &cache_key_fallback v3-angular-node-12-{{ checksum ".bazelversion" }}
+var_4: &cache_key_fallback v4-angular-node-12-{{ checksum ".bazelversion" }}-{{ checksum "yarn.lock" }}
```
Why do we have to include `yarn.lock` into the fallback key?
Couldn't we at least have a second fallback key without `yarn.lock`?

alan-agius4 replied:
I removed but I'd prefer to restore it for the reasons mentioned in https://github.com/angular/angular/pull/41467#discussion_r607807802

alan-agius4 replied:
Should I add the checksum for lock file back or leave it as is?

gkalpak replied:
I am a little concerned that with renovate turned on, `yarn.lock` will change often and result in many cache-misses, which would affect CI times noticeably.

It would be great if there was a way to invalidate the cache after a few days. For example, something like:
`...-{{ checksum ".bazelversion" }}-{{ date +"%y-%m" }}`

Unfortunately, I couldn't find a way to do something like this 😞

alan-agius4 replied:
I think having a large cache will also result in performance overhead.

We can leave it as is and eventually later on.

gkalpak replied:
For reference, I have created #41814 to address this.

gkalpak reviewed (changes requested):
:tada:

IIRC, at some point we had done some testing and found out that caching the `node_modules/` was faster. OOC, have you checked how CI is affected by switching to caching that yarn cache instead?

alan-agius4 committed:
ci: don't save node_modules directory

Storing node_modules directory in cache causes a number of issues, such as increase in the laid out node_modules tree and also in some cases break module resolution. This is because modules are not hoisted correctly when running `yarn install`.

Example of related failures https://app.circleci.com/pipelines/github/angular/angular/30441/workflows/c1a16135-f5c9-478d-a298-d720c6657d52/jobs/952138

A better cache strategy would be to save yarn cache, more information can be found in https://circleci.com/docs/2.0/caching/#yarn-node.

alan-agius4 commented:
>" IIRC, at some point we had done some testing and found out that caching the node_modules/ was faster. OOC, have you checked how CI is affected by switching to caching that yarn cache instead?

Indeed it is slightly slower approx ~40-50s, but by restoring the entire `node_modules` tree isn't the right behaviour, it causes incorrect module hoisting and deduplications which in some cases will results in hard to debug failures such as the one here  https://app.circleci.com/pipelines/github/angular/angular/30441/workflows/c1a16135-f5c9-478d-a298-d720c6657d52/jobs/952138

In the above case the problem was that like this.

We restored `node_modules/foo/node_modules/bar@v1`, when we updated `foo` during `yarn install` `node_modules/foo/node_modules/bar` should have been deleted but wasn't which caused incorrect module resolution.

This is the same reason why we ask users to do the below to solve certain errors when updating one or more packages.
```
rm -rf node_modules
yarn install
```

alan-agius4 committed:
fixup! ci: don't save node_modules directory

gkalpak commented on .circleci/config.yml:
```diff
@@ -187,7 +191,17 @@ commands:
+      # On Windows `~/` is not resolved when using as a CLI value. `../` results in the same path.
+      - run: yarn install --frozen-lockfile --non-interactive --cache-folder ../.cache/yarn
+      # Save dependencies to use on subsequent runs.
+      - save_cache:
+          key: *cache_key_win
+          paths:
+            - ~/.cache/yarn
```
After proposing to move `save_cache` to `setup_win`, I realized that `setup_win` is [not a separate job][1] (unlike [setup][2]) as I thought. So, it probably does not make sense to save the cache in every job that uses `setup_win`.

Therefore, as long as `setup_win` is not a separate job, I think it would be better to move the `save_cache` step back into the `test_win` job 😇 

[1]: https://github.com/angular/angular/blob/aafdad18cd9ec1aef5ea40b276e4541c429299ef/.circleci/config.yml#L183-L186
[2]: https://github.com/angular/angular/blob/aafdad18cd9ec1aef5ea40b276e4541c429299ef/.circleci/config.yml#L224

gkalpak reviewed (approved):
>> OOC, have you checked how CI is affected by switching to caching the yarn cache instead?

> Indeed it is slightly slower but by restoring the entire node_modules tree isn't the right behaviour

Yeah, I am not opposing to the change. Just wondering how CI time will be affected :grin:
A couple of minutes of slow-down is a small time to pay for correctness, I guess :smiley:

<!--
[Note to self] Unresolved comments from previous review:
- https://github.com/angular/angular/pull/41467#discussion_r607818494
-->

alan-agius4 committed:
fixup! ci: add cache for windows run
